### PR TITLE
Add URL input option

### PR DESCRIPTION
**Pull request by aguilerapy**
#42

### CODE REVIEW

File: default/templates/default/to-spreadsheet.html [hunk 1]
```diff
@@ -24,41 +24,92 @@
 </div>
 <div class="response-fail alert alert-danger hidden">
```
Comment (romifz) [inline default/templates/default/to-spreadsheet.html:25]: Is this `div` used?

File: default/static/js/url_uploader.js [hunk 2]
```diff
@@ -0,0 +1,120 @@
+var app = {};
+
+function inputDelete(button) {
```
[the remark this answers is not in the document]
Comment (aguilerapy) [reply default/static/js/url_uploader.js:3]: I create a new snippet (`url_uploader_scripts.html`) where the url_uploader common functions are found.

File: default/templates/default/base-uploader.html [hunk 3]
```diff
@@ -32,41 +36,20 @@
...
-        <span>{% trans "Start" %}</span>
-    </button>
-    </div>
-  </div>
-</form>
+<div class="text-danger">{{ form.type.errors }}</div>
+<div class="btn-group schema-nav" data-toggle="buttons" style="margin-top: 0px;margin-bottom: 15px;">
+    <label class="btn btn-default ">
+        <input type="radio" name="type" class="choice-header toucan-nav-file" autocomplete="off" value="file">
+                {% trans "Upload a file" %}
```
[the remark this answers is not in the document]
Comment (aguilerapy) [reply default/templates/default/base-uploader.html:43]: Done.

File: default/templates/default/base-uploader.html [hunk 4]
```diff
@@ -32,41 +36,20 @@
...
+    </label>
+    <label class="btn btn-default ">
+        <input type="radio" name="type" class="choice-header toucan-nav-input" autocomplete="off" value="input">
+                {% trans "Provide a URL" %}
```
[the remark this answers is not in the document]
Comment (aguilerapy) [reply default/templates/default/base-uploader.html:47]: Done.

File: default/templates/default/combine-packages.html [hunk 5]
```diff
@@ -8,17 +8,29 @@
...
         or <a href="https://standard.open-contracting.org/latest/en/schema/record_package/" target="_blank">record packages</a>
-        to combine them into a single package.
+        into a single package.
     {% endblocktrans %}
-    {% trans "You can also click the <strong>Add a file</strong> link below." %}
 </p>
 <p>
     {% trans "Fill in the <strong>Published date</strong> textbox to set <code>publishedDate</code> in the output package." %}
-    {% trans "Select the correct <strong>Package type</strong> option before uploading." %}
+    {% trans "Select the correct <strong>Package type</strong> option before start." %}
```
[the remark this answers is not in the document]
Comment (aguilerapy) [reply default/templates/default/combine-packages.html:18]: Done.

File: default/templates/default/to-json.html [hunk 6]
```diff
@@ -39,30 +56,65 @@
...
+
+<div role="tabpanel" class="panel-collapse collapse {% if form.type.value == 'input' %}in{% endif %} toucan-nav-input">
+    <div class="input-url-container">
+        <div class="form-group {% if field.errors %}has-error{% endif %}">
+            <label>{% blocktrans trimmed %}
+                Provide the URL to a file:
+                {% endblocktrans %}</label>
+            <input type="text" name="input_url_0" id="id_0'"
+                   class="form-control" value="{{ field.value }}"/>
+            {% if field.errors %}
```
[the remark this answers is not in the document]
Comment (aguilerapy) [reply default/templates/default/to-json.html:78]: Removed from here and from `to-spreadsheet`.

File: default/static/js/url_uploader.js [hunk 7]
```diff
@@ -0,0 +1,120 @@
+var app = {};
```
[the remark this answers is not in the document]
Comment (aguilerapy) [reply default/static/js/url_uploader.js:1]: Done.

File: tests/test_upload_url.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+from django.test import TestCase
+
+
+class UploadUrlTestCase(TestCase):
+    url = 'http://example.com'
+
+    @patch('default.views.os')
+    def test_upload_url(self, mock_os):
+        mock_os.stat = Mock(side_effect=FileNotFoundError())
```
[the remark this answers is not in the document]
Comment (aguilerapy) [reply tests/test_upload_url.py:11]: Done.

File: default/templates/default/to-spreadsheet.html (hunk 1 above)
Comment (aguilerapy) [reply default/templates/default/to-spreadsheet.html:25]: Removed.

File: default/views.py [hunk 9]
```diff
@@ -249,6 +252,43 @@ def perform_to_json(request):
...
+                urlparse(url)
+                with requests.get(url, stream=True) as request_file:
+                    request_file.raise_for_status()
+                    with open(data_file.path, 'wb+') as f:
+                        for chunk in request_file.iter_content(chunk_size=8*1024):
+                            if chunk:
+                                f.write(chunk)
+
+            except (HTTPError, ConnectionError, ValueError, SSLError):
+                return HttpResponse(url, status=400)
```
[the remark this answers is not in the document]
Comment (aguilerapy) [reply default/views.py:280]: Done.

File: tests/test_upload_url.py [hunk 10]
```diff
@@ -0,0 +1,20 @@
+from unittest.mock import Mock, patch
+
+from django.test import TestCase
+
+
+class UploadUrlTestCase(TestCase):
+    url = 'http://example.com'
```
[the remark this answers is not in the document]
Comment (aguilerapy) [reply tests/test_upload_url.py]: Done.

File: default/templates/default/combine-packages.html [hunk 11]
```diff
@@ -8,17 +8,29 @@
...
+        into a single package.
     {% endblocktrans %}
-    {% trans "You can also click the <strong>Add a file</strong> link below." %}
 </p>
 <p>
     {% trans "Fill in the <strong>Published date</strong> textbox to set <code>publishedDate</code> in the output package." %}
-    {% trans "Select the correct <strong>Package type</strong> option before uploading." %}
+    {% trans "Select the correct <strong>Package type</strong> option before starting." %}
 </p>
+<p>
```
[the remark this answers is not in the document]
Comment (aguilerapy) [reply default/templates/default/combine-packages.html:20]: Done.

File: default/templates/default/snippets/url_uploader_scripts.html [hunk 12]
```diff
@@ -0,0 +1,45 @@
+<script type="text/javascript">
```
[the remark this answers is not in the document]
Comment (aguilerapy) [reply default/templates/default/snippets/url_uploader_scripts.html:1]: I think `to.json.html` and `to-spreadsheet.html` use those methods, but they don't extends from `base-uploader.html`

File: default/static/js/uploader.js [hunk 13]
```diff
@@ -171,6 +171,50 @@ var app = {};
         ;
     }
 
+    function performUrlAction() {
```
[the remark this answers is not in the document]
Comment (aguilerapy) [reply default/static/js/uploader.js:174]: Done.

File: default/templates/default/base-uploader.html [hunk 14]
```diff
@@ -23,6 +23,10 @@
     {% trans "You can add more files or remove them, then click on <b>Start</b> again." %}
     {% trans "Files with validation issues will be ignored by default." %}
 </div>
+<div class="response-fail-url alert alert-danger hidden">
```
[the remark this answers is not in the document]
Comment (aguilerapy) [reply default/templates/default/base-uploader.html:26]: Done.

File: default/templates/default/snippets/url_uploader_scripts.html [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+    (function(){
+        this.getCookie = function (name) {
```
[the remark this answers is not in the document]
Comment (aguilerapy) [reply default/templates/default/snippets/url_uploader_scripts.html:3]: Done.

File: locale/es/LC_MESSAGES/django.po [hunk 16]
```diff
@@ -102,48 +105,82 @@ msgid "Files with validation issues will be ignored by default."
...
 msgstr ""
 "Por favor verifique que todos sus archivos son válidos de acuerdo al "
 "estándar OCDS, e intente de nuevo en unos minutos."
 
-#: default/templates/default/base-uploader.html:53
-#: default/templates/default/to-json.html:56
-#: default/templates/default/to-spreadsheet.html:52
+#: default/templates/default/base-uploader.html:43
+msgid "Upload files"
+msgstr "Cargue archivos"
```
[the remark this answers is not in the document]
Comment (aguilerapy) [reply locale/es/LC_MESSAGES/django.po:134]: En varios lugares ya se utiliza "Cargue" para la traducción de "Upload", cambio aquí nada mas o en todos?

File: locale/es/LC_MESSAGES/django.po [hunk 17]
```diff
@@ -156,62 +193,109 @@ msgstr "Combinar Paquetes"
...
+#: default/templates/default/upgrade.html:23
+msgid "Upload a file:"
+msgstr "Cargue un archivo:"
+
+#: default/templates/default/combine-packages.html:26
+#: default/templates/default/compile.html:30
+#: default/templates/default/package-releases.html:24
+#: default/templates/default/upgrade.html:24
+msgid "Drag and drop one or more files."
+msgstr "Arrastre y suelte uno o mas archivos."
```
[the remark this answers is not in the document]
Comment (aguilerapy) [reply locale/es/LC_MESSAGES/django.po:250]: Done.

File: locale/es/LC_MESSAGES/django.po [hunk 18]
```diff
@@ -156,62 +193,109 @@ msgstr "Combinar Paquetes"
...
+
+#: default/templates/default/combine-packages.html:31
+#: default/templates/default/compile.html:35
+#: default/templates/default/package-releases.html:29
+#: default/templates/default/upgrade.html:29
+msgid ""
+"Enter one or more URLs to each file. Click <strong>Add URL</strong> to "
+"provide more URLs."
+msgstr ""
+"Ingresa una o mas URL a cada archivo. Haga click en <strong>Añadir URL</"
```
[the remark this answers is not in the document]
Comment (aguilerapy) [reply locale/es/LC_MESSAGES/django.po:279]: Done.

File: locale/es/LC_MESSAGES/django.po [hunk 19]
```diff
@@ -342,12 +426,14 @@ msgstr "Generar una planilla del esquema"
 #: default/templates/default/index.html:134
 msgid ""
 "You can upload up to 20 files to each tool, and each file should be less "
-"than 20MB. Results remain available for 2 days."
+"than 20MB. Results remain available for 2 days. Another option is to enter "
```
[the remark this answers is not in the document]
Comment (aguilerapy) [reply locale/es/LC_MESSAGES/django.po:429]: Done.

File: locale/es/LC_MESSAGES/django.po [hunk 20]
```diff
@@ -463,27 +538,48 @@ msgid "Published date:"
 msgstr "Fecha de Publicación:"
 
 #: default/templates/default/to-json.html:23
+#: default/templates/default/to-spreadsheet.html:27
 msgid ""
-"Please verify that all your files are valid according to OCDS, and try again "
-"in a few minutes."
+"Please verify that the file in your URL entered are valid OCDS JSON, and try "
```
[the remark this answers is not in the document]
Comment (aguilerapy) [reply locale/es/LC_MESSAGES/django.po:543]: Done.

File: locale/es/LC_MESSAGES/django.po [hunk 21]
```diff
@@ -463,27 +538,48 @@ msgid "Published date:"
...
+"again in a few minutes."
 msgstr ""
-"Por favor verifique que todos sus archivos son válidos de acuerdo al "
+"Por favor verifique que el archivo en la URL ingresada sea válido de acuerdo al "
+"estándar OCDS, e intente de nuevo en unos minutos."
+
+#: default/templates/default/to-json.html:27
+#: default/templates/default/to-spreadsheet.html:31
+msgid ""
+"Please verify that the file are valid OCDS JSON, and try again in a few "
```
[the remark this answers is not in the document]
Comment (aguilerapy) [reply locale/es/LC_MESSAGES/django.po:552]: Done.

File: locale/es/LC_MESSAGES/django.po (hunk 16 above)
[the remark this answers is not in the document]
Comment (romifz) [reply locale/es/LC_MESSAGES/django.po:134]: No sé si es necesario cambiar en todos los casos, podrías cambiar aquí nomás y después vemos.

File: default/templates/default/snippets/url_uploader_scripts.html (hunk 12 above)
[the remark this answers is not in the document]
Comment (romifz) [reply default/templates/default/snippets/url_uploader_scripts.html:1]: I see. There is a lot of code duplication still with those two files. We can work on putting everything in a single javascript file later.

File: locale/es/LC_MESSAGES/django.po (hunk 16 above)
[the remark this answers is not in the document]
Comment (aguilerapy) [reply locale/es/LC_MESSAGES/django.po:134]: Done.

File: default/templates/default/snippets/url_uploader_scripts.html (hunk 12 above)
[the remark this answers is not in the document]
Comment (aguilerapy) [reply default/templates/default/snippets/url_uploader_scripts.html:1]: OK.

**Comment by aguilerapy** [comment]
> Is it possible to verify the format of the file after downloading them to the server?
> 
Done.
> Please make that the "Upload files" option is selected by default.
> 
Done.
> Also, the footer hides the "Start" button after selecting the "Upload files" option:
> 
Done [here](https://github.com/open-contracting/toucan/pull/96).
> It would be nice if we have some kind of visual feedback on how many files are processed and how many are left, instead of the "Processing" modal.
> 
Done.

**Comment by romifz** [comment]
@jpmckinney would you like to review this before merging?

**Comment by jpmckinney** [comment]
I haven't reviewed – looks like you did a thorough review!

I did "Squash and merge" since there were a lot of commits addressing the same issues, which made history a bit unclear.

**Pull request merged**